### PR TITLE
cgroup: move proccesses from root group to new /init group

### PR DESCRIPTION
**Pull request by iwilltry42**
#### Proposed Changes ####

If there are processes running in the root cgroup on a cgroupv2 enabled system, move them to a new `/init` group and enable subtree_control.
For further info, see the discussion and links in the previous PR: https://github.com/k3s-io/k3s/pull/3237

#### Types of Changes ####

Fix for cgroupv2 support.

#### Verification ####

Run on a cgroupv2 enabled system

#### Linked Issues ####

- rancher/k3d#493
- https://github.com/k3s-io/k3s/pull/3237
- https://github.com/moby/moby/commit/ed89041433a031cafc0a0f19cfe573c31688d377

#### Further Comments ####

This is my first time touching cgroups in Go code and reading/writing that part of the k3s code, so there may be some things that could be done different.
The way it's written now, it should only have an effect if there are processes running under the root cgroup.

### CODE REVIEW

**Review by brandond — APPROVED** [review]

File: pkg/daemons/agent/agent.go [hunk 1]
```diff
@@ -199,16 +199,53 @@ func CheckCgroups() (kubeletRoot, runtimeRoot string, hasCFS, hasPIDs bool) {
 	// For Unified (v2) cgroups we can directly check to see what controllers are mounted
 	// under the unified hierarchy.
 	if cgroupsModeV2 {
```
Comment (AkihiroSuda) [inline pkg/daemons/agent/agent.go:201]: Rootless mode already does evacuation by itself: https://github.com/k3s-io/k3s/blob/245efe0d665838dfc228e3442efd11bcd3b39ec8/pkg/rootless/rootless.go#L150

So the evacuation log here should be omitted for rootless mode.
(Or remove the evacuation logic in `rootless.go`)

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

**Comment by AkihiroSuda** [comment]
Can we have CI for cgroup2? (https://github.com/k3s-io/k3s/pull/2887) :pray:

File: pkg/daemons/agent/agent.go (hunk 1 above)
Comment (iwilltry42) [reply pkg/daemons/agent/agent.go:201]: @brandond what do you prefer here?

Comment (brandond) [reply pkg/daemons/agent/agent.go:201]: @AkihiroSuda do you think it makes sense to consolidate the evaction stuff here? If it needs to be done regardless I think we might as well do it all in the agent package and simplify rootless a bit.

Comment (AkihiroSuda) [reply pkg/daemons/agent/agent.go:201]: SGTM

**Comment by brandond** [comment]
> Can we have CI for cgroup2? (#2887) 🙏
 (this is in)

File: pkg/daemons/agent/agent.go (hunk 1 above)
Comment (iwilltry42) [reply pkg/daemons/agent/agent.go:201]: So I'm a bit unsure here.. I feel like my code here could be replaced with `cgrouputil.EvacuateCgroup2("init")` (as it is in https://github.com/k3s-io/k3s/blob/245efe0d665838dfc228e3442efd11bcd3b39ec8/vendor/github.com/rootless-containers/rootlesskit/pkg/parent/cgrouputil/cgrouputil.go#L23), is that correct?
Or do we want to keep it like it is here and just remove this block here: https://github.com/k3s-io/k3s/blob/245efe0d665838dfc228e3442efd11bcd3b39ec8/pkg/rootless/rootless.go#L147-L153 ? :thinking:

Comment (AkihiroSuda) [reply pkg/daemons/agent/agent.go:201]: Yes, you can use cgrouputil.EvacuateCgroup2("init") here, for both rootful and rootless

Comment (iwilltry42) [reply pkg/daemons/agent/agent.go:202]: :heavy_check_mark:

**Comment by brandond** [comment]
Lets see how the tests look, but you need signoff on your commits.

File: .gitignore [hunk 2]
```diff
@@ -29,3 +29,5 @@ __pycache__
 /tests/.vscode
 /sonobuoy-output
 *.tmp
+/.vscode
```
Comment (AkihiroSuda) [inline .gitignore:32]: This should be in your local gitignore

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

**Comment by AkihiroSuda** [comment]
Could you rebase with https://github.com/k3s-io/k3s/pull/3282 to test the change with rootless?

File: pkg/rootless/rootless.go [hunk 3]
```diff
@@ -144,13 +143,6 @@ func createParentOpt(stateDir string) (*parent.Opt, error) {
 	}
 	if selfCgroup2 := selfCgroupMap[""]; selfCgroup2 == "" {
 		logrus.Warnf("enabling cgroup2 is highly recommended, see https://rootlesscontaine.rs/getting-started/common/cgroup2/")
-	} else {
-		selfCgroup2Dir := filepath.Join("/sys/fs/cgroup", selfCgroup2)
-		if unix.Access(selfCgroup2Dir, unix.W_OK) == nil {
-			opt.EvacuateCgroup2 = "k3s_evac"
```
Comment (AkihiroSuda) [inline pkg/rootless/rootless.go:150]: You have to drop this for the child process as well

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

File: pkg/daemons/agent/agent.go [hunk 4]
```diff
@@ -15,6 +15,7 @@ import (
 	"github.com/rancher/k3s/pkg/daemons/executor"
 	"github.com/rancher/k3s/pkg/util"
 	"github.com/rancher/k3s/pkg/version"
+	"github.com/rootless-containers/rootlesskit/pkg/parent/cgrouputil"
```
Comment (AkihiroSuda) [inline pkg/daemons/agent/agent.go:18]: Maybe, we should have a comment line to explain that this logic is not specific to rootless

File: pkg/daemons/agent/agent.go [hunk 5]
```diff
@@ -199,16 +200,33 @@ func CheckCgroups() (kubeletRoot, runtimeRoot string, hasCFS, hasPIDs bool) {
...
-		controllers, err := m.Controllers()
+
+		cgroupRootControllers, err := cgroupRoot.Controllers()
 		if err != nil {
 			return "", "", false, false
 		}
+
+		// evacuate from / to /init
+		if len(cgroupRootProcs) > 0 {
+			if err := cgrouputil.EvacuateCgroup2("init"); err != nil {
```
Comment (AkihiroSuda) [inline pkg/daemons/agent/agent.go:222]: This should be probably called outside of CheckCgroups(). Or rename CheckCgroups(), as this is now more than “checking”

File: pkg/daemons/agent/agent.go (hunk 4 above)
Comment (iwilltry42) [reply pkg/daemons/agent/agent.go:18]: :heavy_check_mark:

File: pkg/rootless/rootless.go (hunk 3 above)
Comment (iwilltry42) [reply pkg/rootless/rootless.go:150]: :heavy_check_mark:

File: pkg/daemons/agent/agent.go (hunk 5 above)
Comment (iwilltry42) [reply pkg/daemons/agent/agent.go]: :heavy_check_mark:

File: pkg/daemons/agent/agent.go [hunk 6]
```diff
@@ -131,7 +131,14 @@ func startKubelet(cfg *config.Agent) error {
 	if err != nil || defaultIP.String() != cfg.NodeIP {
 		argsMap["node-ip"] = cfg.NodeIP
 	}
-	kubeletRoot, runtimeRoot, hasCFS, hasPIDs := CheckCgroups()
+	kubeletRoot, runtimeRoot, hasCFS, hasPIDs, cgroupsModeV2 := CheckCgroups()
+	if cgroupsModeV2 {
```
Comment (AkihiroSuda) [inline pkg/daemons/agent/agent.go:135]: && only when the agent is running as the init

File: pkg/rootless/rootless.go [hunk 7]
```diff
@@ -65,9 +63,6 @@ func Rootless(stateDir string) error {
 	}
 
 	os.Setenv(childEnv, filepath.Join(parentOpt.StateDir, parent.StateFileAPISock))
-	if parentOpt.EvacuateCgroup2 != "" {
-		os.Setenv(evacuateCgroup2Env, "1")
```
Comment (AkihiroSuda) [inline pkg/rootless/rootless.go:69]: `evacuateCgroup2Env = "_K3S_ROOTLESS_EVACUATE_CGROUP2"` on line 26 should be dropped as well

File: pkg/daemons/agent/agent.go (hunk 6 above)
Comment (iwilltry42) [reply pkg/daemons/agent/agent.go:135]: What do you mean by "running as the init"?
Checking if it's PID 1 via `os.GetPid`?

Comment (AkihiroSuda) [reply pkg/daemons/agent/agent.go:135]: That should be fine, but didn't test by myself. Checking the length of /sys/fs/cgroup/cgroup.procs != 0 (as in the previous revision) would be also fine.

File: .gitignore (hunk 2 above)
Comment (iwilltry42) [reply .gitignore]: :heavy_check_mark: 
Didn't even know there was such a thing :D

File: pkg/daemons/agent/agent.go (hunk 6 above)
Comment (iwilltry42) [reply pkg/daemons/agent/agent.go]: :heavy_check_mark:

File: pkg/daemons/agent/agent.go [hunk 8]
```diff
@@ -193,22 +201,33 @@ func addFeatureGate(current, new string) string {
 	return current + "," + new
 }
 
-func CheckCgroups() (kubeletRoot, runtimeRoot string, hasCFS, hasPIDs bool) {
+func CheckCgroups() (kubeletRoot, runtimeRoot string, hasCFS, hasPIDs, v2Evac bool) {
```
Comment (AkihiroSuda) [inline pkg/daemons/agent/agent.go:204]: Now we have lots of ret values, so this function should return a struct.

**Comment by brandond** [comment]
This looks good, but needs a rebase in order to get the CI checks to run.

**Comment by iwilltry42** [comment]
Rebased onto https://github.com/AkihiroSuda/k3s/tree/rootless-ci

**Comment by xiaods** [comment]
need rebase code base

**Comment by iwilltry42** [comment]
@xiaods done :heavy_check_mark: 
Rebasing this was an absolute mess, so I effectively dropped my changes during the rebase and re-applied them on top afterwards.
This PR definitely needs to be squashed and merged, as it now contains pretty useless/dumb commits.

**Comment by dereknola** [comment]
Closing as Out of Date and has been superseded by other PRs.

**Pull request closed without merging**